File: 09/rb/09.rb
#!/usr/bin/env ruby

input = File.readlines('input').map &:to_i

# part 1
val = input.each_cons(26).find{|*a, n| !a.combination(2).map(&:sum).include?(n)}.last
p val

# part 2
start, stop, sum = 0, 0, input[0]
while sum != val
    if sum < val
        stop += 1
        sum += input[stop]
    else
        sum -= input[start]
        start += 1
    end
end
p input[start..stop].minmax.sum

File: 09/hs/09.hs
#!/usr/bin/env runhaskell

import Data.List
import Data.Maybe

part1 input
  | val `elem` sums (take 25 input) = part1 $ tail input
  | otherwise = val
  where val = input !! 25
        sums a = [x+y | (x:xs) <- tails a, y <- xs]

part2 = part2' 0 0
    where part2' sum count input val
            | sum < val = part2' (sum + input!!count) (count+1) input val
            | sum > val = part2' (sum - head input) (count-1) (tail input) val
            | otherwise = (+) <$> minimum <*> maximum $ take count input

main = do
    input <- map read . lines <$> readFile "input"
    let val = part1 input
    putStrLn . show $ val
    putStrLn . show $ part2 input val
